File: archive/llm_written_v1/gemini-2.5-pro/gemm/cpp/gemm_parallel.cpp
#include "gemm_parallel.hpp"
#include <iostream>
#include <vector>
#include <stdexcept>
#include <string>
#include <algorithm>
#include <omp.h>

// Helper functions (mostly unchanged from sequential version)
namespace {

// getSize returns (rows, cols) of a rectangular matrix.
std::pair<int, int> getSize(const Matrix& m) {
    if (m.empty()) return {0, 0};
    return {static_cast<int>(m.size()), static_cast<int>(m[0].size())};
}

// validateMatrix checks that the matrix is non-empty and rectangular.
void validateMatrix(const Matrix& m, const std::string& name) {
    if (m.empty()) {
        throw std::invalid_argument(name + " must have at least one row");
    }
    if (m[0].empty()) {
        throw std::invalid_argument(name + " must have at least one column");
    }
    const int cols = static_cast<int>(m[0].size());
    for (size_t r = 0; r < m.size(); ++r) {
        if (static_cast<int>(m[r].size()) != cols) {
            throw std::invalid_argument(
                name + " is ragged: row " + std::to_string(r) +
                " has " + std::to_string(m[r].size()) +
                " cols, expected " + std::to_string(cols));
        }
    }
}

// generateMatrix returns a rows x cols zero matrix.
Matrix generateMatrix(int rows, int cols) {
    return Matrix(rows, std::vector<double>(cols, 0.0));
}

// partialMatmul computes, for the provided packed blocks:
//   C[m0+i_off][n0+j_off] += alpha * sum_{k=0..kb-1} A[i_off][k] * B[j_off][k]
void partialMatmul(const Matrix& A, const Matrix& B, Matrix& C,
                   double alpha, int m0, int n0, int kb) {
    for (int iOff = 0; iOff < static_cast<int>(A.size()); ++iOff) {
        auto& Ci = C[m0 + iOff];
        const auto& Aik = A[iOff];
        for (int jOff = 0; jOff < static_cast<int>(B.size()); ++jOff) {
            const auto& Bjk = B[jOff];
            double s = 0.0;
            for (int k = 0; k < kb; ++k) {
                s += Aik[k] * Bjk[k];
            }
            Ci[n0 + jOff] += alpha * s;
        }
    }
}

// transpose returns the transpose of a rectangular matrix.
Matrix transpose(const Matrix& m) {
    auto [r, c] = getSize(m);
    Matrix out(c, std::vector<double>(r));
    for (int j = 0; j < c; ++j) {
        for (int i = 0; i < r; ++i) {
            out[j][i] = m[i][j];
        }
    }
    return out;
}

// packMatrix returns a submatrix copy: rows in [b0, b1), cols in [a0, a1).
Matrix packMatrix(const Matrix& matrix, int a0, int a1, int b0, int b1) {
    Matrix out;
    out.reserve(b1 - b0);
    for (int i = b0; i < b1; ++i) {
        const auto& row = matrix.at(i);
        out.emplace_back(row.begin() + a0, row.begin() + a1);
    }
    return out;
}

} // namespace

constexpr int defaultMB = 64;
constexpr int defaultNB = 64;
constexpr int defaultKB = 64;
constexpr int PARALLEL_THRESHOLD_M = 128; // Heuristic threshold

// Gemm computes C := alpha * A * B + beta * C.
// A is (m x k), B is (k x n), C is (m x n).
Matrix gemm(const Matrix& A, const Matrix& B,
            double alpha,
            Matrix* Cptr,
            double beta,
            int MB, int NB, int KB) {
    if (MB == 0) MB = defaultMB;
    if (NB == 0) NB = defaultNB;
    if (KB == 0) KB = defaultKB;

    validateMatrix(A, "A");
    validateMatrix(B, "B");

    auto [m, k] = getSize(A);
    auto [k2, n] = getSize(B);
    if (k != k2) {
        throw std::invalid_argument(
            "shape mismatch: A is (" + std::to_string(m) + "," + std::to_string(k) +
            "), B is (" + std::to_string(k2) + "," + std::to_string(n) + ")");
    }

    Matrix localC;
    Matrix& C = (Cptr ? *Cptr : (localC = generateMatrix(m, n)));

    if (Cptr) {
        validateMatrix(C, "C");
        auto [m2, n2] = getSize(C);
        if (m2 != m || n2 != n) {
            throw std::invalid_argument(
                "C has shape (" + std::to_string(m2) + "," + std::to_string(n2) +
                "); expected (" + std::to_string(m) + "," + std::to_string(n) + ")");
        }
        if (beta != 1.0) {
            #pragma omp parallel for if(m > PARALLEL_THRESHOLD_M)
            for (int i = 0; i < m; ++i) {
                auto& row = C[i];
                for (int j = 0; j < n; ++j) row[j] *= beta;
            }
        }
    }

    if (alpha == 0.0) {
        return C;
    }

    Matrix Bt = transpose(B);

    #pragma omp parallel for collapse(2) if(m * n > PARALLEL_THRESHOLD_M * PARALLEL_THRESHOLD_M)
    for (int n0 = 0; n0 < n; n0 += NB) {
        for (int m0 = 0; m0 < m; m0 += MB) {
            int n1 = std::min(n0 + NB, n);
            int m1 = std::min(m0 + MB, m);
            
            for (int k0 = 0; k0 < k; k0 += KB) {
                int k1 = std::min(k0 + KB, k);
                Matrix Bpack = packMatrix(Bt, k0, k1, n0, n1);
                Matrix Apack = packMatrix(A, k0, k1, m0, m1);
                
                int kb = k1 - k0;
                partialMatmul(Apack, Bpack, C, alpha, m0, n0, kb);
            }
        }
    }

    return C;
}

File: llm_written/claude-sonnet-4-5/gemm/rust/gemm_parallel.rs
use std::sync::{Arc, Mutex};
use std::thread;

pub type Matrix = Vec<Vec<f64>>;

#[derive(Debug)]
pub struct GemmError(pub String);

impl std::fmt::Display for GemmError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for GemmError {}

pub fn get_size(matrix: &Matrix) -> (usize, usize) {
    (matrix.len(), matrix[0].len())
}

pub fn validate_matrix(matrix: &Matrix, name: &str) -> Result<(), GemmError> {
    if matrix.is_empty() {
        return Err(GemmError(format!("{} must be a non-empty array of rows.", name)));
    }
    if matrix[0].is_empty() {
        return Err(GemmError(format!("{} must have at least one column.", name)));
    }
    let cols = matrix[0].len();
    for (r, row) in matrix.iter().enumerate() {
        if row.len() != cols {
            return Err(GemmError(format!(
                "{} is ragged: row {} has {} cols, expected {}.",
                name, r, row.len(), cols
            )));
        }
    }
    Ok(())
}

pub fn zeros(rows: usize, cols: usize) -> Matrix {
    vec![vec![0.0; cols]; rows]
}

pub fn transpose(matrix: &Matrix) -> Matrix {
    let rows = matrix.len();
    let cols = matrix[0].len();
    let mut result = vec![vec![0.0; rows]; cols];
    for j in 0..cols {
        for i in 0..rows {
            result[j][i] = matrix[i][j];
        }
    }
    result
}

pub fn pack_matrix(matrix: &Matrix, a0: usize, a1: usize, b0: usize, b1: usize) -> Matrix {
    let rows = b1 - b0;
    let mut result = Vec::with_capacity(rows);
    for i in 0..rows {
        result.push(matrix[b0 + i][a0..a1].to_vec());
    }
    result
}

fn get_num_threads() -> usize {
    thread::available_parallelism()
        .map(|n| n.get())
        .unwrap_or(4)
}

/// Compute C := alpha * A * B + beta * C
///
/// # Arguments
/// * `a` - m x k matrix
/// * `b` - k x n matrix
/// * `alpha` - Multiplier for A*B
/// * `c` - Optional m x n accumulation buffer (created if None)
/// * `beta` - Multiplier for existing C (applied only if C is provided)
/// * `mb` - Tile size for M dimension
/// * `nb` - Tile size for N dimension
/// * `kb` - Tile size for K dimension
///
/// # Returns
/// m x n result matrix (C)
pub fn gemm(
    a: &Matrix,
    b: &Matrix,
    alpha: f64,
    c: Option<Matrix>,
    beta: f64,
    mb: usize,
    nb: usize,
    kb: usize,
) -> Result<Matrix, GemmError> {
    validate_matrix(a, "A")?;
    validate_matrix(b, "B")?;

    let (m, k) = get_size(a);
    let (k2, n) = get_size(b);

    if k != k2 {
        return Err(GemmError(format!(
            "Shape mismatch: A is ({},{}), B is ({},{}).",
            m, k, k2, n
        )));
    }

    let c = match c {
        None => zeros(m, n),
        Some(mut c_matrix) => {
            validate_matrix(&c_matrix, "C")?;
            let (m2, n2) = get_size(&c_matrix);
            if m2 != m || n2 != n {
                return Err(GemmError(format!(
                    "C has shape ({},{}); expected ({},{}).",
                    m2, n2, m, n
                )));
            }
            if beta != 1.0 {
                for i in 0..m {
                    for j in 0..n {
                        c_matrix[i][j] *= beta;
                    }
                }
            }
            c_matrix
        }
    };

    if alpha == 0.0 {
        return Ok(c);
    }

    // Threshold for using parallel execution (avoid overhead for small matrices)
    const PARALLEL_THRESHOLD: usize = 128;
    let total_work = m * n;
    
    if total_work < PARALLEL_THRESHOLD * PARALLEL_THRESHOLD {
        // Sequential fallback for small matrices
        return gemm_sequential_internal(a, b, alpha, c, m, n, k, mb, nb, kb);
    }

    let bt = transpose(b);

    // Pre-compute all (n0, m0) tile coordinates
    let mut tile_coords = Vec::new();
    let mut n0 = 0;
    while n0 < n {
        let n1 = (n0 + nb).min(n);
        let mut m0 = 0;
        while m0 < m {
            let m1 = (m0 + mb).min(m);
            tile_coords.push((m0, m1, n0, n1));
            m0 += mb;
        }
        n0 += nb;
    }

    // Wrap C in Arc<Mutex> for safe concurrent access
    let c_shared = Arc::new(Mutex::new(c));
    let a_shared = Arc::new(a.clone());
    let bt_shared = Arc::new(bt);
    
    let num_threads = get_num_threads();
    let tiles_per_thread = (tile_coords.len() + num_threads - 1) / num_threads;
    
    let mut handles = vec![];
    
    for thread_id in 0..num_threads {
        let start_idx = thread_id * tiles_per_thread;
        if start_idx >= tile_coords.len() {
            break;
        }
        let end_idx = ((thread_id + 1) * tiles_per_thread).min(tile_coords.len());
        
        let thread_tiles: Vec<_> = tile_coords[start_idx..end_idx].to_vec();
        let a_ref = Arc::clone(&a_shared);
        let bt_ref = Arc::clone(&bt_shared);
        let c_ref = Arc::clone(&c_shared);
        
        let handle = thread::spawn(move || {
            for &(m0, m1, n0, n1) in &thread_tiles {
                // Allocate a local tile buffer for this worker
                let tile_rows = m1 - m0;
                let tile_cols = n1 - n0;
                let mut local_tile = vec![vec![0.0; tile_cols]; tile_rows];

                // Accumulate over k0 dimension sequentially (preserves determinism)
                let mut k0 = 0;
                while k0 < k {
                    let k1 = (k0 + kb).min(k);
                    let kb_len = k1 - k0;

                    let bpack = pack_matrix(&bt_ref, k0, k1, n0, n1);
                    let apack = pack_matrix(&a_ref, k0, k1, m0, m1);

                    // Compute into local tile (no shared state)
                    for (i_off, aik) in apack.iter().enumerate() {
                        for (j_off, bjk) in bpack.iter().enumerate() {
                            let mut s = 0.0;
                            for k_idx in 0..kb_len {
                                s += aik[k_idx] * bjk[k_idx];
                            }
                            local_tile[i_off][j_off] += alpha * s;
                        }
                    }

                    k0 += kb;
                }

                // Write local tile back to C (single lock per tile)
                let mut c_guard = c_ref.lock().unwrap();
                for i in 0..tile_rows {
                    for j in 0..tile_cols {
                        c_guard[m0 + i][n0 + j] += local_tile[i][j];
                    }
                }
            }
        });
        
        handles.push(handle);
    }
    
    // Wait for all threads to complete
    for handle in handles {
        handle.join().unwrap();
    }

    let c = Arc::try_unwrap(c_shared).unwrap().into_inner().unwrap();
    Ok(c)
}

// Sequential fallback for small matrices
fn gemm_sequential_internal(
    a: &Matrix,
    b: &Matrix,
    alpha: f64,
    mut c: Matrix,
    m: usize,
    n: usize,
    k: usize,
    mb: usize,
    nb: usize,
    kb: usize,
) -> Result<Matrix, GemmError> {
    let bt = transpose(b);
    
    let mut n0 = 0;
    while n0 < n {
        let n1 = (n0 + nb).min(n);

        let mut k0 = 0;
        while k0 < k {
            let k1 = (k0 + kb).min(k);
            let bpack = pack_matrix(&bt, k0, k1, n0, n1);

            let mut m0 = 0;
            while m0 < m {
                let m1 = (m0 + mb).min(m);
                let apack = pack_matrix(a, k0, k1, m0, m1);

                let kb_len = k1 - k0;
                
                // Inline partial_matmul
                for (i_off, aik) in apack.iter().enumerate() {
                    let ci = &mut c[m0 + i_off];
                    for (j_off, bjk) in bpack.iter().enumerate() {
                        let mut s = 0.0;
                        for k_idx in 0..kb_len {
                            s += aik[k_idx] * bjk[k_idx];
                        }
                        ci[n0 + j_off] += alpha * s;
                    }
                }

                m0 += mb;
            }
            k0 += kb;
        }
        n0 += nb;
    }

    Ok(c)
}
